synapse/storage/databases/main/devices.py

```diff
@@ -1115,6 +1143,18 @@ async def store_device(
                 if hidden:
                     raise StoreError(400, "The device ID is in use", Codes.FORBIDDEN)
 
+            if auth_provider_id and auth_provider_session_id:
```

Comment:
I don't know if it matters if this should be in a transaction with the `simple_upsert` above or not?

Reply:
there are already multiple things done here not in a transaction, that's why I did not bother

Reply:
I figured that was the rationale. 👍